gems/aws-sdk-core/lib/seahorse/client/plugins/content_type.rb

```diff
@@ -0,0 +1,58 @@
+# frozen_string_literal: true
+
+module Seahorse
+  module Client
+    module Plugins
+      class ContentType < Plugin
+
+        # @api private
+        class Handler < Client::Handler
+          def call(context)
+            body = context.http_request.body
+
+            if (payload = context.operation.input[:payload_member])
+              case payload.shape
+              when Seahorse::Model::Shapes::BlobShape
+                context.http_request.headers['Content-Type'] ||=
+                  'application/octet-stream'
+              when Seahorse::Model::Shapes::StringShape
+                context.http_request.headers['Content-Type'] ||=
+                  'text/plain'
+              else
+                apply_default_content_type(context)
+              end
+            elsif !body.respond_to?(:size) || non_empty_body?(body)
+              apply_default_content_type(context)
+            end
+
+            @handler.call(context)
+          end
+
+          private
+
+          def non_empty_body?(body)
+            body.respond_to?(:size) && body.size.positive?
+          end
+
+          # content-type defaults as noted here:
+          # rest-json: https://smithy.io/2.0/aws/protocols/aws-restxml-protocol.html#content-type
+          # rest-xml: https://smithy.io/2.0/aws/protocols/aws-restxml-protocol.html#content-type
+          def apply_default_content_type(context)
```

Comment:
I don't think this belongs in Seahorse - this is AWS specific logic per protocol.  I'd lean instead towards a generic base class which the protocols can extend and use.